frame/bridge/s2s/issuing/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+use sp_runtime::{DispatchError, DispatchResult};
+use sp_std::vec::Vec;
+// --- darwinia ---
+use darwinia_evm::AddressMapping;
+use darwinia_relay_primitives::{Relay, RelayAccount};
+use dp_asset::token::{Token, TokenInfo};
+use dp_contract::mapping_token_factory::{MappingTokenFactory as mtf, TokenBurnInfo};
+
+const _REGISTERD_ACTION: &[u8] = b"registered(address,address,address)";
+const BURN_ACTION: &[u8] = b"burned(address,address,address,address,uint256)";
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: dvm_ethereum::Config {
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+
+		type ReceiverAccountId: From<[u8; 32]> + Into<Self::AccountId>;
+		type BackingRelay: Relay<
+			RelayOrigin = AccountId<Self>,
+			RelayMessage = (u32, Token, RelayAccount<Self::AccountId>),
+		>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Handle dispatch call from dispatch precompile contract
+		///
+		/// When user burn their tokens, this handler will receive the event from dispatch
+		/// precompile contract, and relay this event to the target chain to unlock asset.
+		#[pallet::weight(0)]
+		pub fn dispatch_handle(origin: OriginFor<T>, input: Vec<u8>) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+
+			// Ensure the input data is long enough
+			ensure!(input.len() >= 8, <Error<T>>::InvalidInput);
+			// Ensure that the user is mapping token factory contract
+			let factory = MappingFactoryAddress::<T>::get();
+			let factory_id = <T as darwinia_evm::Config>::AddressMapping::into_account_id(factory);
+			ensure!(caller == factory_id, <Error<T>>::NotFactoryContract);
+
+			let burn_action = &sha3::Keccak256::digest(&BURN_ACTION)[0..4];
+			if &input[4..8] == burn_action {
+				let burn_info =
+					TokenBurnInfo::decode(&input[8..]).map_err(|_| Error::<T>::InvalidDecoding)?;
+				let recipient = Self::account_id_try_from_bytes(burn_info.recipient.as_slice())?;
+				Self::burn_and_remote_unlock(
+					burn_info.spec_version,
+					burn_info.token_type,
+					burn_info.source,
+					recipient,
+					burn_info.amount,
+				)?;
+			} else {
+				log::trace!("No action match this input selector");
+				return Err(<Error<T>>::InvalidAction.into());
+			}
+			Ok(().into())
+		}
+
+		/// Handle relay message sent from the source backing pallet with relay message
+		///
+		/// If the mapped token is not created, then create it.
+		#[pallet::weight(0)]
+		pub fn cross_receive_and_redeem(
```

Comment:
```suggestion
		pub fn cross_lock_and_issue(
```

It confuse me and says it is redeem  (burn and remote unlock)?

And there is a lazy token creation process here I can not clear say good

Suggest to split registration from cross_lock_and_issue first.

Reply:
The registration will be split from remote_lock_and_issue. I also can not clearly say it better or not. Because it's the native token, and only be registered once.